### PR TITLE
Hacklang property typehint support

### PR DESCRIPTION
**Pull request by mihai-stancu**
Hi,

I have implemented:
- type-hint support for Hack Properties, Getters and Setters;
- type-hint support for PHP7 Getters and Setters;
- refactored PHP5 type-hint support for Setters so that PHP5, PHP7 and Hack all use the same Extractor but with different type-info parsers;

Tested using `phpunit` -- sorry but I don't have experience with `phpspec` yet.

N.B.: PhpDocExtractor needs to be disabled for *.hh files because it expects PHP syntax and things break so I touched that extractor too, to limit its reach on parsing files.

### CODE REVIEW

**Comment by mihai-stancu** [comment]
I should add a GetterExtractor while I'm at it, to sniff return type hints. Since PHP 7 has return type hints on functions this should work nicely for PHP 7 as well as Hack.

**Comment by mihai-stancu** [comment]
Hi,

I have implemented:
* type-hint support for Hack Properties, Getters and Setters;
* type-hint support for PHP7 Getters and Setters;
* refactored PHP5 type-hint support for Setters so that PHP5, PHP7 and Hack all use the same Extractor but with different type-info parsers;

Tested using `phpunit` -- sorry but I don't have experience with `phpspec` yet.

N.B.: PhpDocExtractor needs to be disabled for *.hh files because it expects PHP syntax and things break so I touched that extractor too, to limit its reach on parsing files.

**Comment by mihai-stancu** [comment]
@dunglas bump.

File: src/PropertyInfo/Extractors/GetterExtractor.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+/*
+ * (c) Mihai Stancu <stancu.t.mihai@gmail.com>
```
Comment (dunglas) [inline src/PropertyInfo/Extractors/GetterExtractor.php]: The copyright header must be the one of the project:

```
/*
 * (c) Kévin Dunglas <dunglas@gmail.com>
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */
```

(same for all files)

File: src/PropertyInfo/Extractors/GetterExtractor.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PropertyInfo\Extractors;
+
+use PropertyInfo\Type;
+
+/**
+ * Setter Extractor.
```
Comment (dunglas) [inline src/PropertyInfo/Extractors/GetterExtractor.php]: Should be `Getter Extractor`.

File: src/PropertyInfo/Extractors/PhpDocExtractor.php [hunk 3]
```diff
@@ -102,7 +102,8 @@ public function extractTypes(\ReflectionProperty $reflectionProperty)
      */
     private function getFileReflector(\ReflectionClass $reflectionClass)
     {
-        if (!($fileName = $reflectionClass->getFileName())) {
+        if (!($fileName = $reflectionClass->getFileName())
+        or !in_array(pathinfo($fileName, PATHINFO_EXTENSION), ['php', 'phtml'])) {
```
Comment (dunglas) [inline src/PropertyInfo/Extractors/PhpDocExtractor.php]: Replace `or` by `||` (not the same precedence, can lead to bugs).

It is better to test explicitly on hack type IMO because php files can have a lot of other extensions (`.php3`, `.php4`, `.php5`...).

File: src/PropertyInfo/Tests/Data/HackData.hh [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+namespace PropertyInfo\Tests\Data;
+
+/**
+ * Property Extractor.
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class HackData
+{
+    protected bool $bool;
```
Comment (dunglas) [inline src/PropertyInfo/Tests/Data/HackData.hh]: Can you make properties `private` instead of `protected`.

File: src/PropertyInfo/Tests/Data/HackData.hh [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+ * (c) Mihai Stancu <stancu.t.mihai@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PropertyInfo\Tests\Data;
+
+/**
+ * Property Extractor.
```
Comment (dunglas) [inline src/PropertyInfo/Tests/Data/HackData.hh]: The PHPDoc must be updated to explain what this class do.

File: src/PropertyInfo/Tests/Data/PHP7Data.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+ * (c) Mihai Stancu <stancu.t.mihai@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PropertyInfo\Tests\Data;
+
+/**
+ * Property Extractor.
```
Comment (dunglas) [inline src/PropertyInfo/Tests/Data/PHP7Data.php]: same here

File: src/PropertyInfo/Tests/Data/PHP7Data.php [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class PHP7Data extends PHP5Data
```
Comment (dunglas) [inline src/PropertyInfo/Tests/Data/PHP7Data.php]: Must be Php7Data according to our CS.

File: src/PropertyInfo/Tests/Data/PHP7Data.php [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+{
+    protected $bool;
```
Comment (dunglas) [inline src/PropertyInfo/Tests/Data/PHP7Data.php]: `private` by default.

File: src/PropertyInfo/Tests/ExtractorTest.php [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+require_once(__DIR__.'/../../../vendor/autoload.php');
+
+use PropertyInfo\PropertyInfo;
+
+/**
+ * Property Extractor.
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class ExtractorTest extends \PHPUnit_Framework_TestCase
```
Comment (dunglas) [inline src/PropertyInfo/Tests/ExtractorTest.php]: We use PHPSpec, not PHPUnit. Can you change to it please?

File: src/PropertyInfo/Tests/PHP5DataProvider.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+use PropertyInfo\Extractors\SetterExtractor;
+use PropertyInfo\Tests\Data\PHP5Data;
+
+/**
+ * Property Extractor.
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class PHP5DataProvider
```
Comment (dunglas) [inline src/PropertyInfo/Tests/PHP5DataProvider.php]: Must be `Php5DataProvider`

File: src/PropertyInfo/TypeInfoParser.php [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ */
+
+namespace PropertyInfo;
+
+/**
+ * Property Extractor.
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+interface TypeInfoParser
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParser.php]: `TypeInfoParserInterface`

File: src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+namespace PropertyInfo\TypeInfoParsers;
+
+use PropertyInfo\Type;
+
+/**
+ * Property Extractor.
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class HHVMTypeInfoParser extends NativeTypeInfoParser
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php]: `HhvmTypeInfoParser`

File: src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    /**
+     * @param string $info
+     *
+     * @return array|Type[]
+     */
+    public function parse($info)
+    {
+        $type = new Type();
+
+        if ($info[0] === '?') {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php]: yoda style (`'?' === $info[0]`)

File: src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+            $info = substr($info, 1);
+        }
+
+        if (array_key_exists($info, static::$types)) {
+            $type->setType(static::$types[$info]);
+            $type->setCollection($info === 'array');
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php]: yoda style

File: src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+     */
+    public function parse($info)
+    {
+        $type = new Type();
+
+        if ($info[0] === '?') {
+            $info = substr($info, 1);
+        }
+
+        if (array_key_exists($info, static::$types)) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php]: Prefer `isset` in this case (performance)

File: src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+            $collection = $class->implementsInterface('HH\\Collection');
+
+            $type->setType('object');
+            $type->setClass($info);
+            $type->setCollection($collection);
+
+            return [$type];
+        }
+
+        if (($pos = strpos($info, '<')) !== false) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php]: yoda style

File: src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+            $container = substr($info, 0, $pos);
+            $contents = substr($info, $pos+1, -1);
+            $contents = explode(', ', $contents);
+
+            if ($container === 'array') {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HHVMTypeInfoParser.php]: yoda style

File: src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php [hunk 18]
```diff
@@ -0,0 +1,73 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PropertyInfo\TypeInfoParsers;
+
+use PropertyInfo\TypeInfoParser;
+
+/**
+ * Property Extractor.
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php]: phpdoc

File: src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+     * @param \ReflectionProperty $property
+     *
+     * @return \ReflectionMethod|null
+     */
+    public function getGetter(\ReflectionProperty $property)
+    {
+        $getter = sprintf(static::GETTER_FORMAT, ucfirst($property->getName()));
+        $class = $property->getDeclaringClass();
+        if ($class->hasMethod($getter)) {
+            return $method = $class->getMethod($getter);
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php]: unnecessary assignment (remove `$method =`)

File: src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php [hunk 20]
```diff
@@ -0,0 +1,73 @@
...
+
+    /**
+     * @param \ReflectionProperty $property
+     *
+     * @return \ReflectionParameter|null
+     */
+    public function getSetterParam(\ReflectionProperty $property)
+    {
+        $setter = $this->getSetter($property);
+        if ($setter === null or 1 !== $setter->getNumberOfRequiredParameters()) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php]: yoda style

File: src/PropertyInfo/TypeInfoParsers/PHPTypeInfoParser.php [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PropertyInfo\TypeInfoParsers;
+
+use PropertyInfo\Type;
+
+/**
+ * Property Extractor.
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PHPTypeInfoParser.php]: phpdoc

File: src/PropertyInfo/TypeInfoParsers/PHPTypeInfoParser.php [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class PHPTypeInfoParser extends NativeTypeInfoParser
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PHPTypeInfoParser.php]: `PhpTypeInfoParser`

File: src/PropertyInfo/Extractors/NativeExtractor.php [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+namespace PropertyInfo\Extractors;
+
+use PropertyInfo\TypeExtractorInterface;
+use PropertyInfo\TypeInfoParsers\HHVMTypeInfoParser;
+use PropertyInfo\TypeInfoParsers\PHPTypeInfoParser;
+
+/**
+ * Setter Extractor.
```
Comment (dunglas) [inline src/PropertyInfo/Extractors/NativeExtractor.php]: phpdoc

File: src/PropertyInfo/Extractors/NativeExtractor.php [hunk 24]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+abstract class NativeExtractor implements TypeExtractorInterface
+{
+    /** @var  PHPTypeInfoParser|HHVMTypeInfoParser */
```
Comment (dunglas) [inline src/PropertyInfo/Extractors/NativeExtractor.php]: Should be `TypeInfoParserInterface`

Comment (dunglas) [inline src/PropertyInfo/Extractors/NativeExtractor.php]: And use the multiline style PHPDoc

**Comment by dunglas** [comment]
Sorry I missed the first notification!
Awesome work, it was on my todo list for months.

I made some comments. Can you fix them and add some lines in the doc to explain how to enable PHP7 and HHVM support?

**Comment by mihai-stancu** [comment]
Thanks!

I'll fix most issues today.

On the topic of `phpspec` vs `phpunit` I mentioned that I don't have prior experience with `phpspec` can we merge with the `phpunit` tests and I'll make another PR after I refactor them as `phpspec`?

**Comment by dunglas** [comment]
I prefer to not multiply tools. Another solution is to port the few existing specs to PHPUnit.

File: src/PropertyInfo/Extractors/PhpDocExtractor.php (hunk 3 above)
Comment (mihai-stancu) [inline src/PropertyInfo/Extractors/PhpDocExtractor.php]: I have a personal preference for `and` and `or` specifically because they have lower precedence. It leads to bugs when you're not aware of it but for conformity's sake I'll flip them the other way round.

`*.hh` files are one example but PhpDoc fails to parse `*.php` files which have PHP7 syntax inside -- what about those? I could name those files `*.php7` for now and exclude them as such...

File: src/PropertyInfo/Tests/Data/HackData.hh (hunk 5 above)
Comment (mihai-stancu) [inline src/PropertyInfo/Tests/Data/HackData.hh]: This class doesn't do anything it's just a dummy class used for the tests -- like `SetterDummy`. But I have updated the doc comment nonetheless.

**Comment by mihai-stancu** [comment]
I commited the fixes but wasn't able to test the beforehand -- I have PHP7 compiled on another laptop -- hope they still run :-).

**Comment by dunglas** [comment]
Travis has tested it for us!

**Comment by dunglas** [comment]
Your previous comment seems to have been deleted. Strange.

File: src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+    /**
+     * @param \ReflectionProperty $property
+     *
+     * @return string|null
+     */
+    public function getGetterReturnType(\ReflectionProperty $property)
+    {
+        $getter = $this->getGetter($property);
+
+        if (PHP_VERSION >= 7 and $getter !== null and $getter->hasReturnType()) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php]: `and` here too. The problem is precisely that many developers don't know the precedence trick and it can lead to bugs.
`and` and `or` are forbidden in the Symfony CS (and we follow Symfony CS) and are reported as violations by most quality checking tools including SensioLabs Insight and Scrutinizr (and we use both).

File: src/PropertyInfo/Extractors/SetterExtractor.php [hunk 26]
```diff
@@ -10,48 +10,23 @@
 namespace PropertyInfo\Extractors;
```
Comment (dunglas) [inline src/PropertyInfo/Extractors/SetterExtractor.php:10]: Missing license header.

**Comment by dunglas** [comment]
Tests fail with PHP7: https://travis-ci.org/dunglas/php-property-info/jobs/70364772

**Comment by mihai-stancu** [comment]
Travis runs the `phpspec` not `phpunit`.

File: src/PropertyInfo/Extractors/SetterExtractor.php (hunk 26 above)
Comment (mihai-stancu) [inline src/PropertyInfo/Extractors/SetterExtractor.php:10]: I don't see the missing header in my files.

Comment (dunglas) [inline src/PropertyInfo/Extractors/SetterExtractor.php:10]: Sorry my bad it is here.

**Comment by mihai-stancu** [comment]
Does the PhpDocExtractor read doc comments off of Properties as well as Getters and Setters?

It would stand to reason to refactor the PhpDocExtractor to use the TypeInfoParser logic as well.

**Comment by dunglas** [comment]
would be cool

**Comment by dunglas** [comment]
@mihai-stancu any news about that PR?

**Comment by mihai-stancu** [comment]
I checked my `phpunit` tests yesterday, they worked correctly on all of my configs (php5, php7, hhvm). 

The `phpspec` tests were failing on an issue related to mock objects -- something about receiving an object which does not respect the expected interface -- I think it may be cause by returning null instead of empty array in some instances of my implementation.

I'll transcribe the `phpspec` tests into `phpunit` tomorrow and make sure those work.

**Comment by mihai-stancu** [comment]
The `phpspec` tests from `master` fail with the same error when ran with PHP7. Which means the error wasn't induced by my PR branch.

I'll just go on with rewriting the tests as PHPUnit if that's alright with you.

**Comment by dunglas** [comment]
Strange, it doesn't fail for others PR such as #2.
IMO you porting PHPUnit tests to PHPSpec will be easier (PHPSpec also use Prophecy).

File: src/PropertyInfo/Tests/ExtractorTest.php [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+                    $this->assertEquals($expectedType, $actualType->getType());
+                } else {
+                    $this->assertEquals('float', $actualType->getType());
+                }
+
+                if (is_string($expectedClass)) {
+                    $this->assertEquals($expectedClass, $actualType->getClass());
+                }
+            } else {
+                $this->fail('Using ' . get_class($extractor) . ' type "' . $expectedType . '" resulted in an empty type list');
```
Comment (dunglas) [inline src/PropertyInfo/Tests/ExtractorTest.php]: Use `sprintf` for such cases to improve readability.

**Comment by dunglas** [comment]
For the data provider part: https://github.com/coduo/phpspec-data-provider-extension

**Comment by dunglas** [comment]
Ok it only fails with PHP7. And it seems that PHPSpec isn't supporting PHP7 for now so it's OK.

**Comment by mihai-stancu** [comment]
I've rewritten the phpspec tests as phpunit tests and added `phpunit` to `travis` & `composer` to be run automatically. PHP7 obviously still fails for `phpspec` but works fine on `phpunit`.

**Comment by dunglas** [comment]
Can you fix new issues raised by Scrutinizr please? https://scrutinizer-ci.com/g/dunglas/php-property-info/inspections/c440da42-b348-429e-9ad1-04efa97435c8

**Comment by mihai-stancu** [comment]
Hey,

I looked them over a bit, but didn't have the time to fix them. I'll get back to them tomorrow.

**Comment by mihai-stancu** [comment]
I fixed the two big issues in Scrutinizer, the rest are either tests related or small.

**Comment by mihai-stancu** [comment]
Bump.

**Comment by dunglas** [comment]
Can you squash your commits?

**Comment by mihai-stancu** [comment]
Squashed.

**Comment by dunglas** [comment]
Thank you @mihai-stancu

**Comment by dunglas** [comment]
@mihai-stancu your commit introduced a major Insight violations and some minors: https://insight.sensiolabs.com/projects/29b845cf-106d-45b4-99af-271f2dc3f7d5/analyses/23

Can you fix them please?

File: composer.json [hunk 28]
```diff
@@ -23,11 +23,17 @@
       "PropertyInfo\\": "src/PropertyInfo"
     }
   },
+  "autoload-dev": {
```
Comment (dunglas) [inline composer.json:26]: Why is this necessary?

Comment (soullivaneuh) [inline composer.json:26]: It's for tests classes namespace definition AFAIK.

File: phpunit.xml [hunk 29]
```diff
@@ -0,0 +1,5 @@
+<phpunit
```
Comment (dunglas) [inline phpunit.xml:1]: Must be `phpunit.xml.dist`: https://github.com/dunglas/DunglasApiBundle/blob/master/phpunit.xml.dist

File: composer.json (hunk 28 above)
Comment (soullivaneuh) [inline composer.json:26]: Tests classes structure is not the same as source. It should be, isn't it?

Comment (dunglas) [inline composer.json:26]: Yes it must be.

File: src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * @param string $info
+     *
+     * @return Type|null
+     */
+    public function parseSimpleTypes($info)
+    {
+        $type = new Type();
+
+        if (isset(static::$types[$info])) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php:98]: the succession of `if`/`else` blocks can be avoided using `return` statements.

File: src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php [hunk 31]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * @param string $info
+     *
+     * @return Type|null
+     */
+    protected function parseContainerTypes($info)
+    {
+        $type = new Type();
+
+        if (false !== ($pos = strpos($info, '<'))) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php:124]: Same here, better use `return` than `if`/`else` to reduce the complexity.

File: src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php [hunk 32]
```diff
@@ -0,0 +1,76 @@
...
+
+    /**
+     * @param \ReflectionProperty $property
+     *
+     * @return \ReflectionParameter|null
+     */
+    public function getSetterParam(\ReflectionProperty $property)
+    {
+        $setter = $this->getSetter($property);
+        if (null === $setter or 1 !== $setter->getNumberOfRequiredParameters()) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php:63]: Please no `or`

Comment (soullivaneuh) [inline src/PropertyInfo/TypeInfoParsers/NativeTypeInfoParser.php:63]: See #3 for this.

File: src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php [hunk 33]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * @param \ReflectionProperty $property
+     *
+     * @return string|null
+     */
+    public function getGetterReturnType(\ReflectionProperty $property)
+    {
+        $getter = $this->getGetter($property);
+
+        if (PHP_VERSION >= 7 && $getter !== null && $getter->hasReturnType()) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:44]: Please use yoda style (according to Symfony CS): `null === $getter`.

File: src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php [hunk 34]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+    /**
+     * @param \ReflectionProperty $property
+     *
+     * @return string|null
+     */
+    public function getGetterReturnType(\ReflectionProperty $property)
+    {
+        $method = $this->getGetter($property);
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/HhvmTypeInfoParser.php:55]: Potential bug: `getGetter` can return `null`.

File: src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    /**
+     * @param \ReflectionProperty $property
+     *
+     * @return string|null
+     */
+    public function getGetterReturnType(\ReflectionProperty $property)
+    {
+        $getter = $this->getGetter($property);
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php]: Potential bug: `getGetter` can return `null`.

File: src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php [hunk 36]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * @param \ReflectionProperty $property
+     *
+     * @return string|null
+     */
+    public function getSetterParamType(\ReflectionProperty $property)
+    {
+        $param = $this->getSetterParam($property);
+        if (null !== $param) {
+            if (PHP_VERSION >= 7) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:58]: Reduce the complexity of this method by removing `elseif` and `else` statements (`return` are sufficient)

File: src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php [hunk 37]
```diff
@@ -0,0 +1,100 @@
...
+    public function parse($info)
+    {
+        $type = new Type();
+
+        $collection = false;
+        $class = null;
+
+        if ($info === null) {
+            return null;
+        } elseif ($info === 'double') {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:85]: Same here, the `elseif` if not necessary because of the  `return` on the previous line.

File: src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php [hunk 38]
```diff
@@ -0,0 +1,100 @@
...
+     * @return array|Type[]|null
+     */
+    public function parse($info)
+    {
+        $type = new Type();
+
+        $collection = false;
+        $class = null;
+
+        if ($info === null) {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:83]: Yoda style

File: src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php [hunk 39]
```diff
@@ -0,0 +1,100 @@
...
+            return null;
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:84]: `return` (`null` is not necessary)

File: src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php [hunk 40]
```diff
@@ -0,0 +1,100 @@
...
+        } elseif ($info === 'double') {
+            $info = 'float';
+        } elseif ($info === 'array') {
```
Comment (dunglas) [inline src/PropertyInfo/TypeInfoParsers/PhpTypeInfoParser.php:87]: yoda style

File: .travis.yml [hunk 41]
```diff
@@ -28,3 +27,4 @@ install:
 
 script:
   - vendor/bin/phpspec run
+  - vendor/bin/phpunit tests
```
Comment (dunglas) [inline .travis.yml:30]: `vendor/bin/phpunit` must be sufficient is the config is OK.

File: tests/DoctrineExtractors/DoctrineExtractorTest.php [hunk 42]
```diff
@@ -0,0 +1,109 @@
...
+use PropertyInfo\TypeExtractorInterface;
+
+/**
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class DoctrineExtractorTest extends \PHPUnit_Framework_TestCase
+{
+    public function extractorsDataProvider()
+    {
+        $properties = array(
```
Comment (dunglas) [inline tests/DoctrineExtractors/DoctrineExtractorTest.php:27]: Short array syntax.

File: tests/DoctrineExtractors/DoctrineExtractorTest.php [hunk 43]
```diff
@@ -0,0 +1,109 @@
...
+            array(
```
Comment (dunglas) [inline tests/DoctrineExtractors/DoctrineExtractorTest.php:28]: Short array syntax.

File: tests/DoctrineExtractors/DoctrineExtractorTest.php [hunk 44]
```diff
@@ -0,0 +1,109 @@
...
+                'name'       => 'id',
+                'type'       => 'int',
+                'collection' => false,
+                'class'      => null,
+            ),
+            array(
```
Comment (dunglas) [inline tests/DoctrineExtractors/DoctrineExtractorTest.php:34]: same here

File: tests/DoctrineExtractors/DoctrineExtractorTest.php [hunk 45]
```diff
@@ -0,0 +1,109 @@
...
+                'name'       => 'guid',
+                'type'       => 'string',
+                'collection' => false,
+                'class'      => null,
+            ),
+            array(
```
Comment (dunglas) [inline tests/DoctrineExtractors/DoctrineExtractorTest.php:40]: same here

File: tests/DoctrineExtractors/DoctrineExtractorTest.php [hunk 46]
```diff
@@ -0,0 +1,109 @@
...
+                'name'       => 'bool',
+                'type'       => 'bool',
+                'collection' => false,
+                'class'      => null,
+            ),
+            array('name' => 'binary', 'type' => 'resource', 'collection' => false, 'class' => null),
```
Comment (dunglas) [inline tests/DoctrineExtractors/DoctrineExtractorTest.php:46]: same here and below

File: tests/NativeExtractors/Data/HackData.hh [hunk 47]
```diff
@@ -0,0 +1,340 @@
...
+ *
+ * @author Mihai Stancu <stancu.t.mihai@gmail.com>
+ */
+class HackData
+{
+    private bool $bool = true;
+    private int $int = 0;
+    private float $float = 0.0;
+    private string $string = '';
+    private array $array = array();
```
Comment (dunglas) [inline tests/NativeExtractors/Data/HackData.hh:25]: Short array syntax here and bellow.

File: tests/DoctrineExtractors/DoctrineExtractorTest.php (hunk 42 above)
Comment (soullivaneuh) [inline tests/DoctrineExtractors/DoctrineExtractorTest.php:27]: Will be auto-corrected with #4.

**Comment by dunglas** [comment]
Sorry @mihai-stancu my review was incomplete. Can you fix issues raised by Insight and my new comments please (you can open another PR or reopen this one).

**Comment by mihai-stancu** [comment]
OK sure.

File: composer.json (hunk 28 above)
Comment (mihai-stancu) [inline composer.json:26]: The purpose is to load the Tests namespace only for dev but it can be omitted.

**Comment by mihai-stancu** [comment]
Should we remove the `phpspec`s from the repo so as not to use two tools for the same purpose?

File: composer.json (hunk 28 above)
Comment (dunglas) [inline composer.json:26]: Ok please remove it.

**Comment by dunglas** [comment]
Yes `phpspec` can now be removed.

File: .travis.yml (hunk 41 above)
Comment (mihai-stancu) [inline .travis.yml:30]: If not configured in phpunit.xml then PHPUnit expects a `phpunit [options] <directory>`. I'll add the directory to the XML.

Comment (dunglas) [inline .travis.yml:30]: Yes, it's why you need a `phpunit.xml.dist` file (and the `.dist` allow custom configuration not versioned if necessary).

Comment (mihai-stancu) [inline .travis.yml]: Apparently  even if the testsuites are configured in the xml it still needs me to specify the directory (ex.: `.`).

**Comment by mihai-stancu** [comment]
@Soullivaneuh I'm getting a `PHP Fatal error:  Class 'Doctrine\Tests\Common\Cache\CacheTest'` I assume it's due recent changes. Any idea where it's coming from?

**Comment by mihai-stancu** [comment]
Done. PR #5
